src/modules/mtev_kafka.cpp

```diff
@@ -0,0 +1,428 @@
+      else {
+        consumer_group_string = consumer_group;
+        free(consumer_group);
+      }
+      try {
+        auto conn = std::make_unique<kafka_connection>(host_string, port, topic_string,
+                                                       consumer_group_string);
+        _conns.push_back(std::move(conn));
+      }
+      catch (std::exception &exception) {
+        mtevL(nlerr, "ERROR: Couldn't connect to %s:%d, topic %s (Exception: %s) - skipping\n",
+              host_string.c_str(), port, topic_string.c_str(), exception.what());
+      }
+      catch (...) {
+        mtevL(nlerr, "ERROR: Couldn't connect to %s:%d, topic %s (unknown exception) - skipping\n",
+              host_string.c_str(), port, topic_string.c_str());
+      }
+    }
+    mtev_conf_release_sections_read(mqs, number_of_conns);
+  }
+  ~kafka_module_config() = default;
+  void set_poll_timeout(const std::chrono::milliseconds poll_timeout)
+  {
+    _poll_timeout = poll_timeout;
+  }
+  void set_poll_limit(const int32_t poll_limit) { _poll_limit = poll_limit; }
+  int poll()
+  {
+    for (const auto &conn : _conns) {
+      int32_t per_conn_cnt = 0;
+      rd_kafka_message_t *msg = nullptr;
+      while (
+        (_poll_limit == 0 || per_conn_cnt < _poll_limit) &&
+        (nullptr != (msg = rd_kafka_consumer_poll(conn->rd_consumer, _poll_timeout.count())))) {
+        conn->stats.msgs_in++;
+        per_conn_cnt++;
+        if (msg->err == RD_KAFKA_RESP_ERR_NO_ERROR) {
+          mtev_rd_kafka_message_t *m = mtev_rd_kafka_message_alloc(msg, mtev_rd_kafka_message_free);
+          mtev_kafka_handle_message_dyn_hook_invoke(m);
+          mtev_rd_kafka_message_deref(m);
+        }
+        else {
+          mtevL(nlerr, "ERROR: Got error reading from %s, topic %s: %s\n",
+                conn->broker_with_port.c_str(), conn->topic.c_str(), rd_kafka_err2str(msg->err));
+          conn->stats.errors++;
+          rd_kafka_message_destroy(msg);
+        }
+      }
+    }
+    return 0;
+  }
+  int show_console(const mtev_console_closure_t &ncct)
+  {
+    for (const auto &conn : _conns) {
+      conn->write_to_console(ncct);
+    }
+    return 0;
+  }
+
+private:
+  std::vector<std::unique_ptr<kafka_connection>> _conns;
+  std::chrono::milliseconds _poll_timeout;
+  int32_t _poll_limit;
+};
+
+static kafka_module_config *the_conf = nullptr;
+
+static kafka_module_config *get_or_load_config(mtev_dso_generic_t *self)
+{
+  if (the_conf) {
+    return the_conf;
+  }
+  the_conf = static_cast<kafka_module_config *>(mtev_image_get_userdata(&self->hdr));
+  if (the_conf) {
+    return the_conf;
+  }
+  the_conf = new kafka_module_config{};
+  mtev_image_set_userdata(&self->hdr, the_conf);
+  return the_conf;
+}
+
+static int kafka_logio_open(mtev_log_stream_t ls)
+{
+  (void) ls;
+  return 0;
+}
+
+static int
+  kafka_logio_write(mtev_log_stream_t ls, const struct timeval *whence, const void *buf, size_t len)
+{
+  // TODO: Need a write function
+  return len;
```

Comment:
This should return -1 until an implementation exists.

Reply:
Done